hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py

```diff
@@ -0,0 +1,241 @@
+        of the wsi. If the objective power is not found in the wsi properties, the base_objective_power is used
+        instead. If the objective power is not found in the wsi properties and base_objective_power is None, an error
+        is raised., defaults to None
+        :param replace_ampersand_by: A string that is used to replace ampersands in the src file name, defaults to
+        UNDERSCORE. This is useful because ampersands in file names can cause problems in cloud storage.
+        :param compression: The compression that is used to save the tiff file, defaults to COMPRESSION.ADOBE_DEFLATE
+        aka ZLIB that is lossless compression.
+        :param tile_size: The size of the tiles that are used to write the tiff file, defaults to 512.
+        :param backend: The backend that is used to read the wsi file, defaults to WSIBackend. OPENSLIDE which is
+        compatible with most wsi formats.
+        """
+        self.output_folder = output_folder
+        self.image_key = image_key
+        self.src_format = src_format
+        if target_magnifications is not None:
+            target_magnifications.sort(reverse=True)
+        self.target_magnifications = target_magnifications if target_magnifications else None
+        self.add_lowest_magnification = add_lowest_magnification
+        self.replace_ampersand_by = replace_ampersand_by
+        self.base_objective_power = base_objective_power
+        self.wsi_reader = WSIReader(backendbackend=backend)
+        self.compression = compression
+        self.tile_size = tile_size
+
+    def get_tiff_path(self, src_path: Path) -> Path:
+        """Returns the path to the tiff file that will be created from the src file. The tiff file is saved in the
+        output_folder with the same name as the src file but with the tiff extension. Ampersands are replaced by the
+        replace_ampersand_by string.
+
+        :param src_path: The path to the src file.
+        :return: The path to the tiff file that will be created from the src file.
+        """
+        tiff_filename = src_path.name.replace(self.src_format, WSIFormat.TIFF)
+        tiff_filename = tiff_filename.replace(AMPERSAND, self.replace_ampersand_by)
+        return self.output_folder / tiff_filename
+
+    def _get_base_objective_power(self, wsi_obj: OpenSlide) -> float:
+        """Returns the objective power of the wsi. The objective power is extracted from the wsi properties. If the
+        objective power is not found in the wsi properties, the base_objective_power is used instead.
+
+        :param wsi_obj: The wsi object in openslide format
+        :raises ValueError: Raises an error if the objective power is not found in the wsi properties and
+        base_objective_power is None
+        :return: The base objective power of the wsi
+        """
+        base_objective_power = wsi_obj.properties.get(self.OBJECTIVE_POWER_KEY, self.base_objective_power)
```

Comment:
I'm confused. YOu're reading that out with a key that is `openslide....` does that exist if the backend is `cucim`?

Reply:
That is why we have default_base_objectiove_power to use if that field is not part of the properties. 
I haven't been able to find the right tag to set for base objective power for tiff files, https://www.loc.gov/preservation/digital/formats/content/tiff_tags.shtml#:~:text=TIFF%20files%20declare%20and%20describe%20their%20content%20by,or%20another%20image%20compression%20option%20has%20been%20used.
Here's the available metadata using cucim 
![image](https://user-images.githubusercontent.com/37396332/222115640-ef632fdb-af8b-43d7-a074-986018773e83.png)
This OBJECTIVE_POWER_KEY is only available in the original ndpi files. 

Now that you said that I haven't tested the cpode with a different backend and I think it won't work given that all wsi_obj are typed as OpenSlide

I just made this change (making backend flexible) but I think we better stick with Openslide

Reply:
Agree, let's stick with openslide only for now.